### PR TITLE
Integrate AgentMarket.outcome_token_pool into betting strategies

### PR DESCRIPTION
**Pull request by evangriffiths**
- Adding a 'full' kelly bet based using the olas code
- Using this in the KellyBettingStrategy if supported
- Refactoring the market-moving-bet function to not be OmenMarket-specific

### CODE REVIEW

File: prediction_market_agent_tooling/deploy/betting_strategy.py [hunk 1]
```diff
@@ -168,11 +170,26 @@ def calculate_trades(
         market: AgentMarket,
     ) -> list[Trade]:
         adjusted_bet_amount = self.adjust_bet_amount(existing_position, market)
-        kelly_bet = get_kelly_bet(
-            adjusted_bet_amount,
-            market.current_p_yes,
-            answer.p_yes,
-            answer.confidence,
+        # TODO use market.get_outcome_str_from_bool after https://github.com/gnosis/prediction-market-agent-tooling/pull/387 merges
```
Comment (kongzii) [inline prediction_market_agent_tooling/deploy/betting_strategy.py:173]:
```suggestion
```

**Review by kongzii — APPROVED** [review]

**Comment by evangriffiths** [comment]
Trying to verify the calculation in ``, and it looks like there's a bug somewhere. You would expect the kelly bet to asymptote to the market-moving bet, but see from the plots (for two different markets) that this isn't the case:

1.
Screenshot 2024-09-10 at 01 09 22

2.
Screenshot 2024-09-10 at 01 09 53

To reproduce:

```
import numpy as np
import pytest
from matplotlib import pyplot as plt
from prediction_market_agent_tooling.markets.agent_market import FilterBy, SortBy
from prediction_market_agent_tooling.markets.omen.omen import OmenAgentMarket
from prediction_market_agent_tooling.tools.betting_strategies.kelly_criterion import (
    get_kelly_bet_full,
)
from prediction_market_agent_tooling.tools.betting_strategies.market_moving import (
    _sanity_check_omen_market_moving_bet,
    get_market_moving_bet,
)

market = OmenAgentMarket.get_binary_markets(
    limit=1,
    sort_by=SortBy.CLOSING_SOONEST,
    filter_by=FilterBy.OPEN,
)[0]

yes_outcome_pool_size = market.outcome_token_pool[
    market.get_outcome_str_from_bool(True)
]
no_outcome_pool_size = market.outcome_token_pool[
    market.get_outcome_str_from_bool(False)
]
max_bets = np.linspace(0.0, 200, 100)

# Define a list of colors for different estimated_p_yes values
colors = plt.cm.viridis(np.linspace(0, 1, len([0.05, 0.42, 0.71])))

for i, estimated_p_yes in enumerate([0.57, 0.74, 0.92]):
    kelly_bets = []
    market_moving_bets = []
    market_moving_bet = get_market_moving_bet(
        yes_outcome_pool_size=yes_outcome_pool_size,
        no_outcome_pool_size=no_outcome_pool_size,
        market_p_yes=market.current_p_yes,
        target_p_yes=estimated_p_yes,
        fee=market.fee,
    )
    _sanity_check_omen_market_moving_bet(market_moving_bet, market, estimated_p_yes)

    market_moving_bet_signed = (
        market_moving_bet.size
        if market_moving_bet.direction == True
        else -market_moving_bet.size
    )

    for max_bet in max_bets:
        kelly_bet = get_kelly_bet_full(
            yes_outcome_pool_size=yes_outcome_pool_size,
            no_outcome_pool_size=no_outcome_pool_size,
            estimated_p_yes=estimated_p_yes,
            confidence=1.0,
            max_bet=max_bet,
        )
        kelly_bet_signed = (
            kelly_bet.size if kelly_bet.direction == True else -kelly_bet.size
        )
        kelly_bets.append(kelly_bet_signed)
        market_moving_bets.append(market_moving_bet_signed)

        # TODO kelly bet is not moving the market as expected!
        if max_bet == max_bets[-1]:
            with pytest.raises(ValueError) as e:
                _sanity_check_omen_market_moving_bet(kelly_bet, market, estimated_p_yes)
            assert e.match("Bet does not move market to target_p_yes")

    plt.plot(
        max_bets,
        kelly_bets,
        label=f"Kelly bet, {estimated_p_yes=:.2f}",
        color=colors[i],
        linestyle="-",
    )
    plt.plot(
        max_bets,
        market_moving_bets,
        label=f"Market-moving bet, target_p_yes={estimated_p_yes:.2f}",
        color=colors[i],
        linestyle="--",
    )

plt.xlabel(f"Max bet (xDai)")
plt.ylabel("Kelly bet size (xDai)")
plt.title(
    f"Market-moving vs. kelly bet, market_p_yes={market.current_p_yes:.2f}, pool_size={(yes_outcome_pool_size+no_outcome_pool_size):.2f}"
)
plt.legend(bbox_to_anchor=(1.05, 1), loc="upper left")
plt.tight_layout()
plt.show()
```

File: prediction_market_agent_tooling/tools/betting_strategies/kelly_criterion.py [hunk 2]
```diff
@@ -57,4 +52,86 @@ def get_kelly_bet(
...
+                    - 4 * b * x**2 * y * p * c
+                    + 4 * b * x * y**2
+                )
+            )
+        )
+        ** (1 / 2)
+    )
+    denominator = 2 * (x**2 * f - y**2 * f)
+    if denominator == 0:
+        return SimpleBet(direction=True, size=0)
```
[the remark this answers is not in the document]
Comment (evangriffiths) [reply prediction_market_agent_tooling/tools/betting_strategies/kelly_criterion.py:134]: Yes, but that doesn't seem like desirable behaviour! Have made an issue to fix: https://github.com/gnosis/prediction-market-agent-tooling/issues/401

**Comment by evangriffiths** [comment]
Script to reproduce

Screenshot 2024-09-12 at 01 36 00

on this PR https://github.com/gnosis/prediction-market-agent-tooling/pull/389

**Pull request merged**